jdaviz/core/marks.py

```diff
@@ -233,23 +347,82 @@ def __init__(self, shadowing, shadow_width=1, **kwargs):
 
     def _on_shadowing_changed(self, change):
-        attr = change['name']
-        if attr[0] in ['stroke_width', 'marker_size']:
-            value = change['new'] + self._shadow_width if change['new'] else 0
-        elif attr not in self._sync_traits:
+        super()._on_shadowing_changed(change)
+        if change['name'] in ['stroke_width', 'marker_size']:
+            # apply the same, but increased by the shadow width
+            setattr(self, change['name'],
+                    change['new'] + self._shadow_width if change['new'] else 0)
+        return
+
+
+class ShadowLabelFixedY(Label, ShadowMixin):
+    """
+    Label whose position shadows that of a parent ``shadowing`` line and will flip alignment
+    based on whether it is left or right of the center of the viewer.
+    """
+    def __init__(self, viewer, shadowing, shadow_traits=['visible'],
+                 y=0.95, point_index=0, **kwargs):
+        super().__init__(**kwargs)
+        self._viewer = viewer
+        self.y = [y]
+        self.scales['y'] = LinearScale(min=0, max=1)
+        self._point_index = point_index
+        self._setup_shadowing(shadowing,
+                              shadow_traits,
+                              ['x', 'scales', 'labels', 'colors'])
+
+        viewer.state.add_callback("x_min", lambda x_min: self._update_align())
+        viewer.state.add_callback("x_max", lambda x_max: self._update_align())
+
+    def _force_redraw(self):
+        # TODO: bug in bqplot that change in align/colors traitlet doesn't update immediately,
+        # we'll get around it in the meantime by just forcing the Label to see a change to the
+        # text traitlet
+        text = self.text
+        self.text = ['']
+        self.text = text
+
+    def _update_align(self):
+        if not isinstance(self.scales.get('x'), LinearScale):
             return
+        # determine alignment automatically
+        if self.scales['x'].min == 0 and self.scales['x'].max == 1:
+            # then we're in axes units, so just check position compared to 0.5
+            is_to_right = self.x[0] > 0.5
         else:
-            value = change['new']
+            # then we're in data units, so check position compared to the median of the axes limits
+            is_to_right = self.x[0] > (self._viewer.state.x_min + self._viewer.state.x_max) / 2.
+
+        if is_to_right and self.align != 'end':
+            self.align = 'end'
+            # force redraw by re-updating label
+            self._force_redraw()
+        if not is_to_right and self.align != 'start':
+            self.align = 'start'
+            # force redraw by re-updating label
+            self._force_redraw()
 
-        setattr(self, attr, value)
+    def _on_shadowing_changed(self, change):
+        super()._on_shadowing_changed(change)
+        if change['name'] == 'labels':
+            self.text = [change['new'][self._point_index]]
+        elif change['name'] in ['x', 'colors']:
```

Comment:
Total nitpick: I prefer tuple over list when the thing is immutable.
```suggestion
        elif change['name'] in ('x', 'colors'):
```